### PR TITLE
callback: add tests

### PR DESCRIPTION
**Pull request by jacquese**
#### Tests
How did you test these changes prior to submitting this pull request?
What automated tests are included in this PR?

ran them, and tests for callbacks (don't pass yet because the functionality will change to match the test)

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp [hunk 1]
```diff
@@ -0,0 +1,155 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Tests/TestDefinitions.h"
+
+#include "SpatialView/Callbacks.h"
+
+#define CALLBACKS_TEST(TestName) GDK_TEST(Core, Callbacks, TestName)
+
+using namespace SpatialGDK;
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp:9]: Nit - even in cpp files I'd prefer to not have this. Slightly better is using specific types rather than the whole namespace.

However for tests why not just put them in the same namespace?

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	// WHEN
+	Callbacks.Invoke(1);
+
+	// THEN
+	TestTrue("Callback was invoked once", InvokeCount == 1);
+
+	return true;
+}
+
+CALLBACKS_TEST(GIVEN_Callbacks_With_Callback_WHEN_Callback_Removes_Other_Callback_THEN_Calls_Both_Callbacks)
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp:133]: Same as the above. Might be worth testing that after the second invocation that the second callback is removed.

The other edge cases are

1. The same callback is added and removed during callback invocation.

1. Nested calls to Invoke.

I guess the fanciest edge case of all is, from a single callback, adding another callback, calling invoke, removing that callback during the nested invocation. I doubt this is worth explicitly testing. 

Maybe we should make it check that callback depth is 0 when calling invoke. Then nested invocations aren't a thing.

**Review by samiwh — APPROVED** [review]
Generally lgtm. Left what are mostly nits or things that should probably be fixed in the underlying feature rather than the tests.

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp (hunk 1 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp:9]: the namespace thing was copied for the worker view test, which I based this off- although I see why it is an issue. Are you saying it should instead all be defined inside the GDK namespace, or all types should be prefixed?

Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp:9]: changed to specify for each type as it felt nicer

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	// THEN
+	TestFalse("Callbacks is no longer empty", IsEmpty);
+
+	return true;
+}
+
+CALLBACKS_TEST(GIVEN_Callbacks_With_Callback_WHEN_Invoke_Called_THEN_Invokes_Callback)
+{
+	// GIVEN
+	TCallbacks<int> Callbacks;
```
[the remark this answers is not in the document]
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp:33]: it was to just use one callback type for the whole test suite. If the void form is a special case that needs testing separately, happy to use that for the tests where it makes sense.

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	Callbacks.Remove(Id);
+	Callbacks.Invoke(1);
+
+	// THEN
+	TestTrue("Callback was invoked once", InvokeCount == 1);
+
+	return true;
+}
+
+CALLBACKS_TEST(GIVEN_Callbacks_With_Callback_WHEN_Callback_Adds_Other_Callback_THEN_Only_Calls_First_Callback)
```
[the remark this answers is not in the document]
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp:116]: I think you're right. Left it out because then it felt like this test was doing two things, but it's worth covering

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp:33]: spoke offline, back to int

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp:116]: added

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp (hunk 2 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp]: nested invocations no longer allowed, so that is fine

added test to check that on second invocation only one is called

same callback added and removed: This one is a bit weirder. Added a test which asserts that the callback added and removed is never called.

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+	return true;
+}
+
+CALLBACKS_TEST(GIVEN_Callbacks_With_Callback_WHEN_Invoke_Called_THEN_Invokes_Callback)
+{
+	// GIVEN
+	TCallbacks<int> Callbacks;
+	bool Invoked = false;
+	const FIntCallback Callback = [&Invoked](int _){Invoked = true;};
```
[the remark this answers is not in the document]
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/CallbacksTest.cpp]: fixed
